### PR TITLE
Big QP parameter refactor; validate param :types for Cards

### PR DESCRIPTION
**Pull request by camsaul**
Part of https://github.com/metabase/metaboat/issues/144

The main purpose of this PR is to enforce a new constraint that all `:parameters` passed in to Card query endpoints match up with the template tag parameters for the Card's query (only native queries can have template tag parameters). This means you can no longer pass arbitrary MBQL parameters when running a Card query. Parameters must match existing parameters and have allowed types, e.g. you cannot pass a `string/!=` parameter when the template tag specifies a `:widget-type` of `:string/=` (for security reasons explained in the comments in the PR).

As part of this project I undertook a few important refactors:

- Code for running queries in the context of a specific Card or Dashboard+Card has been moved to new namespaces `metabase.query-processor.card` and `metabase.query-processor.dashboard` respectively. Previously they were in `metabase.api.card` and `metabase.api.dashboard`.
- Schemas for native query `:template-tags` and for `:parameters` are now finally moved into `metabase.mbql.schema` itself. This actually involved consolidating dispersed information from four or five different namespaces. I tried my best to document all the possible values of these
- (As part of the schema above) finally coming up with a list of the allowed parameter types and widget types -- previously this list didn't exist anywhere. I had to update some tests because they were doing things that didn't make sense or were missing some keys they should have been passing
- The backend parameters code was kinda fragile and every time I changed something, something else would break in an e2e test. I added close to 600 LoC of new backend tests as part of this PR

### CODE REVIEW

File: test/metabase/api/card_test.clj [hunk 1]
```diff
@@ -11,17 +11,28 @@
             [metabase.api.pivots :as pivots]
             [metabase.driver.sql-jdbc.execute :as sql-jdbc.execute]
             [metabase.http-client :as http]
-            [metabase.models :refer [Card CardFavorite Collection Dashboard Database ModerationReview
-                                     Pulse PulseCard PulseChannel PulseChannelRecipient Table ViewLog]]
+            [metabase.models
```
Comment (dpsutton) [inline test/metabase/api/card_test.clj:14]: This change is probably not desired. Auto format got a bit hungry

File: test/metabase/pulse/render/body_test.clj [hunk 2]
```diff
@@ -7,6 +7,12 @@
             [metabase.pulse.render.test-util :as render.tu]
             [schema.core :as s]))
 
+(use-fixtures :each
```
Comment (dpsutton) [inline test/metabase/pulse/render/body_test.clj:10]: this is weird. Did this not get merged already?

Comment (dpsutton) [reply test/metabase/pulse/render/body_test.clj:10]: oh i see it was added to js-svg-test and copied to body-test here

**Review by dpsutton — APPROVED** [review]
This is fantastic. Nice!

File: test/metabase/api/card_test.clj (hunk 1 above)
Comment (camsaul) [reply test/metabase/api/card_test.clj:14]: good catch. I'll fix

File: test/metabase/pulse/render/body_test.clj (hunk 2 above)
Comment (camsaul) [reply test/metabase/pulse/render/body_test.clj:10]: yeah, I ran into it again in this namespace as well

**Comment by camsaul** [comment]
Presto JDBC failure was unrelated and fixed upstream by #19256. Everything else is green -- merging

**Pull request merged**